Architecture.md

```diff
@@ -0,0 +1,165 @@
+
+## Bindings to Python Objects
+[`src/types`] contains bindings to [built-in types](https://docs.python.org/3/library/stdtypes.html)
+of Python, such as `dict` and `list`.
+Due to historical reasons, Python's `object` is called `PyAny` and placed in [`src/types/any.rs`].
+Currently, `PyAny` is a straight-forward wrapper of `ffi::PyObject`, like:
+```rust
+#[repr(transparent)]
+pub struct PyAny(UnsafeCell<ffi::PyObject>);
+```
+
+All built-in types are defined as a C struct.
+For example, `dict` is defined as:
+```c
+typedef struct {
+    /* Base object */
+    PyObject ob_base;
+    /* Number of items in the dictionary */
+    Py_ssize_t ma_used;
+    /* Dictionary version */
+    uint64_t ma_version_tag;
+    PyDictKeysObject *ma_keys;
+    PyObject **ma_values;
+} PyDictObject;
+```
+
+However, we cannot access such a specific data structure with `#[cfg(Py_LIMITED_API)]` set.
+Thus, all builtin objects are implemented as opaque types by wrapping `PyAny`, like:
+```rust
+#[repr(transparent)]
+pub struct PyDict(PyAny);
+```
+
+Note that `PyAny` is not a pointer, and it is usually used as a pointer to the object in the
+Python heap, as `&PyAny`.
+This design choice can be changed
+(see the discussion in [#1056](https://github.com/PyO3/pyo3/issues/1056)).
+
+Since we need lots of boilerplate for implementing common traits for these types
+(e.g., `AsPyPointer`, `AsRef<PyAny>`, and `Debug`), we have some macros in
+[`src/types/mod.rs`].
+
+## PyClass
+[`src/pycell.rs`], [`src/pyclass.rs`], and [`src/type_object.rs`] contains types and
+traits to make `#[pyclass]` work.
+Also, [`src/pyclass_init.rs`] and [`src/pyclass_slots.rs`] have related functionalities.
+
+To realize object-oriented programming in C, all Python objects must have the following two fields
+at the beginning.
+```rust
+#[repr(C)]
+pub struct PyObject {
+    pub ob_refcnt: usize,
+    pub ob_type: *mut PyTypeObject,
+    ...
+}
+```
+Thanks to this guarantee, casting `*mut A` to `*mut PyObject` is valid if `A` is a Python object.
+
+To ensure this guarantee, we have a wrapper struct `PyCell<T>` in [`src/pycell.rs`] which is roughly:
+```rust
+#[repr(C)]
+pub struct PyCell<T: PyClass> {
+    object: crate::ffi::PyObject,
+    inner: T,
+}
+```
+Thus, when copying a Rust struct to a Python object, we first allocate `PyCell` on the Python heap and then
+copy `T`.
+Also, `PyCell` provides [RefCell](https://doc.rust-lang.org/std/cell/struct.RefCell.html)-like methods
+to ensure Rust's borrow rules.
+See [the document](https://docs.rs/pyo3/latest/pyo3/pycell/struct.PyCell.html) for more.
+
+`PyCell<T>` requires that `T` implements `PyClass`.
+This trait is somewhat complex and derives many traits, but the most important one is `PyTypeObject`
+in [`src/type_object.rs`].
+`PyTypeObject` is also implemented for built-in types.
+Type objects are singletons, and all Python types have their unique type objects.
```

Comment:
It could be worth noting [PEP 630](https://www.python.org/dev/peps/pep-0630/) - we're technically out-of-date by still having singleton type objects. But I doubt we're fixing this any time soon 😅